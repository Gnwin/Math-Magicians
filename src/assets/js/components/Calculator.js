import React from 'react';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div className="calculator">
        <div className="answer">0</div>
        <div>AC</div>
        <div>+-</div>
        <div>%</div>
        <div>/</div>
        <div>7</div>
        <div>8</div>
        <div>9</div>
        <div>x</div>
        <div>4</div>
        <div>5</div>
        <div>6</div>
        <div>-</div>
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>+</div>
        <div className="zero">0</div>
        <div>.</div>
        <div>=</div>
      </div>
    );
  }
}

export default Calculator;
