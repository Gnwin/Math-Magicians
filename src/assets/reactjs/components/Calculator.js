/* eslint-disable */
import React from 'react';
import PropTypes from 'prop-types';

import Screen from './Screen';
import calculate from './calculate';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      total: '',
      next: '',
      operation: '',
    };
  }

  componentDidMount = () => {
    const buttons = document.querySelectorAll('.center');
    buttons.forEach((button) => {
      button.addEventListener('click', this.clickButton);
    });
  }

  clickButton = (event) => {
    event.preventDefault();
    const buttonContent = event.target.innerHTML;
    const operands = this.props.operands;
    let calcData;
    if (operands.includes(buttonContent)) {
      calcData = calculate(this.state, buttonContent);
      this.setState({
        total: calcData.total,
        next: '',
        operation: calcData.operation,
      });
    } else if (buttonContent === '=') {
      calcData = calculate(this.state, buttonContent);
      this.setState({
        total: calcData.total,
        next: '',
        operation: '',
      });
    } else if (buttonContent === 'AC') {
      calcData = calculate(this.state, buttonContent);
      this.setState({
        total: '',
        next: '',
        operation: '',
      });
    } else if (buttonContent === '+/-') {
      calcData = calculate(this.state, buttonContent);
      this.setState({
        total: 'undefined',
        next: '',
        operation: '',
      });
    } else {
      calcData = calculate(this.state, buttonContent);
      this.setState((oldState) => ({
        next: oldState.next + buttonContent,
      }));
    }
  }

  render() {
    return (
      <div className="calculator">
        <Screen data={this.state} />
        <div className="center">AC</div>
        <div className="center">+/-</div>
        <div className="center">%</div>
        <div className="center orange">รท</div>
        <div className="center">7</div>
        <div className="center">8</div>
        <div className="center">9</div>
        <div className="center orange">x</div>
        <div className="center">4</div>
        <div className="center">5</div>
        <div className="center">6</div>
        <div className="center orange">-</div>
        <div className="center">1</div>
        <div className="center">2</div>
        <div className="center">3</div>
        <div className="center orange">+</div>
        <div className="center zero">0</div>
        <div className="center">.</div>
        <div className="center orange">=</div>
      </div>
    );
  }
}

export default Calculator;

Calculator.propTypes = {
  numbers: PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
  operands: PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
};

/* eslint-enable */
