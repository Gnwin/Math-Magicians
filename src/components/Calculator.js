import React from 'react';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: 0,
    };
  }

  render() {
    const { value } = this.state;
    return (
      <div>{value}</div>
    );
  }
}

export default Calculator;
