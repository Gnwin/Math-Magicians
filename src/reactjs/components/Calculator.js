/* eslint-disable */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import Screen from './Screen';
import calculate from './calculate';

const Calculator = (props) => {
  const [calcState, setCalcState] = useState({
    total: '',
    next: '',
    operation: '',
  });
  useEffect(() => {
    const buttons = document.querySelectorAll('.center');
    buttons.forEach((button) => {
      button.addEventListener('click', clickButton);
    });
  }, []);
 
  const clickButton = (event) => {
    event.preventDefault();
    const buttonContent = event.target.innerHTML;
    if (props.operands.includes(buttonContent)) {
      setCalcState((oldCalcState) => ({
        total: (calculate(oldCalcState, buttonContent)).total,
        next: '',
        operation: (calculate(oldCalcState, buttonContent)).operation
      }))
    } else if (buttonContent === '=') {
      setCalcState((oldCalcState) => ({
        total: (calculate(oldCalcState, buttonContent)).total,
        next: '',
        operation: ''
      }))
    } else if (buttonContent === 'AC') {
      setCalcState({
        total: '',
        next: '',
        operation: '',
      })
    } else if (buttonContent === '+/-') {  
      setCalcState((oldCalcState) => ({
        total: calculate(oldCalcState, buttonContent).total,
        next: calculate(oldCalcState, buttonContent).next,
        operation: calculate(oldCalcState, buttonContent).operation,
      }));
    } else if (buttonContent === '.') {      
      setCalcState((oldCalcState) => ({
        total: oldCalcState.total ? (calculate(oldCalcState, buttonContent)).total : '',
        next: oldCalcState.next ? (calculate(oldCalcState, buttonContent)).next : '',
        operation: oldCalcState.operation ? (calculate(oldCalcState, buttonContent)).operation : '',
      }));
    } else {
      setCalcState((oldCalcState) => ({
        total: oldCalcState.total ? (calculate(oldCalcState, buttonContent)).total : '',
        next: oldCalcState.next + buttonContent,
        operation: oldCalcState.operation ? (calculate(oldCalcState, buttonContent)).operation : '',
      }));
    }
  }

  return (
    <div className="calculator">
      <Screen data={calcState} />
      <div className="center">AC</div>
      <div className="center">+/-</div>
      <div className="center">%</div>
      <div className="center orange">รท</div>
      <div className="center">7</div>
      <div className="center">8</div>
      <div className="center">9</div>
      <div className="center orange">x</div>
      <div className="center">4</div>
      <div className="center">5</div>
      <div className="center">6</div>
      <div className="center orange">-</div>
      <div className="center">1</div>
      <div className="center">2</div>
      <div className="center">3</div>
      <div className="center orange">+</div>
      <div className="center zero">0</div>
      <div className="center">.</div>
      <div className="center orange">=</div>
    </div>
  );

}

export default Calculator;

Calculator.propTypes = {
  numbers: PropTypes.array.isRequired,
  operands: PropTypes.array.isRequired,
};

/* eslint-enable */
