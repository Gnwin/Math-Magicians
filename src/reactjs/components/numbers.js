const limit = 10;
const numbers = (num) => [...Array(num).keys()];

export default numbers(limit);
