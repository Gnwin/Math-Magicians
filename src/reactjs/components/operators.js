const operators = ['x', '-', '+', 'รท', '+/-', '%'];

export default operators;
