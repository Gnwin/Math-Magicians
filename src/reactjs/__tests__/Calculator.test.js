// import react-testing methods
import {render, fireEvent, waitFor, screen, cleanup} from '@testing-library/react'

import renderer from 'react-test-renderer';

// import { renderHook } from '@testing-library/react-hooks';

// add custom jest matchers from jest-dom
import '@testing-library/jest-dom'

// import { renderHook } from '@testing-library/react-hooks'

import numbers from '../components/numbers';
import operators from '../components/operators';
import Calculator from '../components/Calculator';


describe('it tests the Calculator Component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Calculator operators={operators} numbers={numbers} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the Calculator Component', () => {
    render(<Calculator operators={operators} numbers={numbers} />);
    const calcComponent = screen.getByTestId('calc');
    expect(calcComponent).toBeInTheDocument();
  });
});