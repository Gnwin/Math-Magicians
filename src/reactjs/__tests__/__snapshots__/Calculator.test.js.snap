// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it tests the Calculator Component renders correctly 1`] = `
<div
  className="calculator"
  data-testid="calc"
>
  <div
    className="answer"
  >
    0
  </div>
  <div
    className="center"
  >
    AC
  </div>
  <div
    className="center"
  >
    +/-
  </div>
  <div
    className="center"
  >
    %
  </div>
  <div
    className="center orange"
  >
    รท
  </div>
  <div
    className="center"
  >
    7
  </div>
  <div
    className="center"
  >
    8
  </div>
  <div
    className="center"
  >
    9
  </div>
  <div
    className="center orange"
  >
    x
  </div>
  <div
    className="center"
  >
    4
  </div>
  <div
    className="center"
  >
    5
  </div>
  <div
    className="center"
  >
    6
  </div>
  <div
    className="center orange"
  >
    -
  </div>
  <div
    className="center"
  >
    1
  </div>
  <div
    className="center"
  >
    2
  </div>
  <div
    className="center"
  >
    3
  </div>
  <div
    className="center orange"
  >
    +
  </div>
  <div
    className="center zero"
  >
    0
  </div>
  <div
    className="center"
  >
    .
  </div>
  <div
    className="center orange"
  >
    =
  </div>
</div>
`;
